### PR TITLE
Add upgrade tests into the e2e-tests-latest-eventing

### LINKED ISSUE #94: Add the upgrade tests
Add upgrade tests into the existing prow job 
```
pull-knative-eventing-operator-integration-tests-on-latest-eventing
```

The major reason is that this prow job has already tested whether the operator work with the latest generate manifest of knative eventing. Compared to creation of new prow, we can reuse this guy to do more.

### PR DESCRIPTION
**Pull request by houshengbo**
# Issue to be fixed

Fixes #94 

## Proposed Changes

* This PR adds the tests to verify the correct number and names of knative eventing
deployments.

* The test tag postupgrade is added, marking the tests to run after upgrade to the
latest HEAD of operator, with the latest generated manifest of knative eventing.

## Release Note

```release-note
```

### CODE REVIEW

File: test/common/common.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline test/common/common.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: test/upgrade/knativeeventingupgrade_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+//// +build postupgrade
+
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline test/upgrade/knativeeventingupgrade_test.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: test/common/common.go (hunk 1 above)
Comment (mattmoor) [inline test/common/common.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

File: test/upgrade/knativeeventingupgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
+// +build postupgrade
+
+/*
+Copyright 2019 The Knative Authors
```
Comment (mattmoor) [inline test/upgrade/knativeeventingupgrade_test.go]: Incorrect copyright year:
```suggestion
Copyright 2020 The Knative Authors
```

**Comment by aliok** [comment]
cc @cardil you might want to have a look perhaps?

**Comment by cardil** [comment]
Doesn't this upgrade test should mirror the knative/eventing#2388 ? Now, it only runs some simple test after upgrade.

In knative/eventing#2388 I'm doing upgrade & downgrade, while validating that running knative services (probe), will still run, and we will not lose **any** events during upgrade & downgrade.

**Comment by houshengbo** [comment]
@cardil That's the goal as well, so far we add some tests to verify the deployments of the knatvie eventing in a macro scope from the operator's perspective. We will ultamately own the tests in eventing with the tag preupgrade, and postupgrade.
Also, we also move on with serving-operator to add the upgrade flow as well, so all the upgrade tests will live in operators in future.

File: test/e2e-tests-latest-eventing.sh [hunk 4]
```diff
@@ -95,12 +95,14 @@ function generate_latest_eventing_manifest() {
...
+TIMEOUT=20m
+
 # If we got this far, the operator installed Knative Eventing
 header "Running tests for Knative Eventing Operator"
 failed=0
 
-# Run the integration tests
-go_test_e2e -timeout=20m ./test/e2e || failed=1
+# Run the postupgrade tests
+go_test_e2e -tags=postupgrade -timeout=${TIMEOUT} ./test/upgrade || failed=1
```
Comment (aliok) [inline test/e2e-tests-latest-eventing.sh]: +1 on this. Also good that we're changing the name of this file with https://github.com/knative/eventing-operator/pull/96

**Comment by aliok** [comment]
@houshengbo while working on this, I found some errors in the .sh file and fixed them.
Here's the PR that is sent against your PR branch: https://github.com/houshengbo/eventing-operator/pull/1

File: test/upgrade/knativeeventingupgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+		common.KnativeEventingVerify(t, clients, names)
+		common.KnativeEventingDelete(t, clients, names)
+	})
+}
+
+// knativeEventingVerifyDeployment verify whether the deployments have the correct number and names.
+func knativeEventingVerifyDeployment(t *testing.T, clients *test.Clients, names test.ResourceNames) {
+	// Knative Eventing has 4 deployments.
+	expectedNumDeployments := 4
+	deploys := []string{"eventing-controller", "eventing-webhook", "imc-controller", "imc-dispatcher"}
```
Comment (aliok) [inline test/upgrade/knativeeventingupgrade_test.go]: After https://github.com/knative/eventing/pull/2519 this is changed

```
// Knative Eventing has 5 deployments.
expectedNumDeployments := 5
deploys := []string{"eventing-controller", "broker-controller", "eventing-webhook", "imc-controller", "imc-dispatcher"}
```

**Comment by aliok** [comment]
@houshengbo 

Added some comments.

I think let's merge this before waiting on my PR (https://github.com/knative/eventing-operator/pull/95).

About @cardil 's comments: let's do them later in sync with @cardil 

However, I have some doubts about the purpose of this exercise.
* `e2e-tests.sh` is using the release manifest in `kodata`. So:
  * it starts the operator
  * installs Knative eventing using the 0.12 manifest
  * doing some tests (like deployment recreated when deleted etc.)
* `e2e-tests-latest-eventing.sh` which uses the manifest from master of Knative eventing.
  *  it starts the operator
  * installs Knative eventing using the manifest from master
  * doing some tests like (checking number of deployments and expected deployments etc.)

As we're going test the updade of the *operator*, we need the resources created by a previous version of the operator. Simply pointing the operator to old manifest won't work because in https://github.com/knative/eventing-operator/pull/95 we're deleting obsolote resources by name.

So, IMO, the process should be:
* Deploy the latest released version of operator
* Create CR, so that old Knative eventing version is created (with obsolote resources)
* Delete the operator and deploy the operator from local code
* Check expected resources and new resources (obsolote things should be gone, new things should be there)

IMO, let's merge https://github.com/knative/eventing-operator/pull/95 first. Then adapt this PR.

**Comment by houshengbo** [comment]
I have renamed the script from `e2e-tests-latest-eventing.sh` to `e2e-upgrade-tests.sh`, which is dedicated to run the upgrade tests for operator, including the operator itself and the knative. 

Prow job has been created to run the upgrade tests as well in test-infra.

File: test/upgrade/knativeeventingupgrade_test.go (hunk 5 above)
Comment (houshengbo) [reply test/upgrade/knativeeventingupgrade_test.go]: done.

File: test/client/setup.go [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Knative Authors
+Copyright 2020 The Knative Authors
```
Comment (k4leung4) [inline test/client/setup.go]: in the future, this doesnt need to changed
the copyright year is based on when the file is first added.

File: test/e2e-upgrade-tests.sh [hunk 7]
```diff
@@ -48,9 +51,37 @@ function install_eventing_operator() {
   wait_until_pods_running default || fail_test "Eventing Operator did not come up"
 }
 
+function install_latest_operator_release() {
+  header "Installing Knative Eventing operator latest public release"
+  local full_url="https://github.com/knative/eventing-operator/releases/download/${LATEST_EVENTING_RELEASE_VERSION}/eventing-operator.yaml"
+
+  local RELEASE_YAML="$(mktemp)"
```
Comment (k4leung4) [inline test/e2e-upgrade-tests.sh]: lower case
release_yaml

File: test/e2e/knativeeventingdeployment_test.go [hunk 8]
```diff
@@ -16,17 +16,9 @@ limitations under the License.
 package e2e
 
 import (
-	"errors"
-	"path/filepath"
-	"runtime"
+	"knative.dev/eventing-operator/test/client"
```
Comment (k4leung4) [inline test/e2e/knativeeventingdeployment_test.go]: group with other knative.dev/eventing-operator imports

**Comment by aliok** [comment]
I finally had time to review this PR after the changes.

Great work @houshengbo , thank you!

**Pull request merged**